File: bms/go/geo/geo_r2_Rect_Intersection/geo_r2_Rect_Intersection.go
package test

import (
	"math"
)

func (r geo_r2_Rect) Intersection(other geo_r2_Rect) geo_r2_Rect {
	xx := r.X.Intersection(other.X)
	yy := r.Y.Intersection(other.Y)
	if xx.IsEmpty() || yy.IsEmpty() {
		return geo_r2_EmptyRect()
	}

	return geo_r2_Rect{xx, yy}
}

func (i geo_r1_Interval) Intersection(j geo_r1_Interval) geo_r1_Interval {
	// Empty intervals do not need to be special-cased.
	return geo_r1_Interval{
		Lo: math.Max(i.Lo, j.Lo),
		Hi: math.Min(i.Hi, j.Hi),
	}
}

func (i geo_r1_Interval) IsEmpty() bool { return i.Lo > i.Hi }

func geo_r2_EmptyRect() geo_r2_Rect {
	return geo_r2_Rect{geo_r1_EmptyInterval(), geo_r1_EmptyInterval()}
}

func geo_r1_EmptyInterval() geo_r1_Interval { return geo_r1_Interval{1, 0} }

type geo_r2_Rect struct {
	X, Y geo_r1_Interval
}

type geo_r1_Interval struct {
	Lo, Hi float64
}

File: translations_only/success/go/geo/geo_r2_Rect_Intersection/claude2/Hinted/geo_r2_Rect_Intersection.rs

use std::f64::{self, MAX, MIN}; 

struct GeoR1Interval {
    lo: f64,
    hi: f64,
}

struct GeoR2Rect {
    x: GeoR1Interval,
    y: GeoR1Interval, 
}

fn intersection(r: &GeoR2Rect, other: &GeoR2Rect) -> GeoR2Rect {
    let xx = intersect_intervals(&r.x, &other.x);
    let yy = intersect_intervals(&r.y, &other.y);
    if is_empty(&xx) || is_empty(&yy) {
        empty_rect()
    } else {
        GeoR2Rect { x: xx, y: yy }
    }
}

fn intersect_intervals(i: &GeoR1Interval, j: &GeoR1Interval) -> GeoR1Interval {
    GeoR1Interval {
        lo: f64::max(i.lo, j.lo),
        hi: f64::min(i.hi, j.hi),
    }
}

fn is_empty(i: &GeoR1Interval) -> bool {
    i.lo > i.hi
}

fn empty_rect() -> GeoR2Rect {
    GeoR2Rect {
        x: empty_interval(),
        y: empty_interval(), 
    }
}

fn empty_interval() -> GeoR1Interval {
    GeoR1Interval { lo: 1.0, hi: 0.0 }  
}

